### PR TITLE
feat(cache): remove temporary cache after filesystem scanning

### PR DESCRIPTION
**Pull request by afdesk**
## Description
We generate pseudo layer id (artifact id and blob id) for filesystem, but there is no reason to keep them in the cache. The cache cannot be used next time since it is pseudo id. They should be removed after filesystem scanning.

## Checklist
- [x] I've read the [guidelines for contributing](https://github.com/aquasecurity/trivy/blob/main/CONTRIBUTING.md) to this repository.
- [x] I've followed the [conventions](https://github.com/aquasecurity/trivy/blob/main/CONTRIBUTING.md#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/scanner/scan.go [hunk 1]
```diff
@@ -105,7 +106,7 @@ func NewScanner(driver Driver, ar artifact.Artifact) Scanner {
 }
 
 // ScanArtifact scans the artifacts and returns results
-func (s Scanner) ScanArtifact(ctx context.Context, options types.ScanOptions) (types.Report, error) {
+func (s Scanner) ScanArtifact(ctx context.Context, options types.ScanOptions, ac cache.ArtifactCache) (types.Report, error) {
```
Comment (knqyf263) [inline pkg/scanner/scan.go:109]: `s.artifact` already has cache. We don't have to pass it here.

File: pkg/scanner/scan.go [hunk 2]
```diff
@@ -121,6 +122,12 @@ func (s Scanner) ScanArtifact(ctx context.Context, options types.ScanOptions) (t
 		log.Logger.Warnf("The vulnerability detection may be insufficient because security updates are not provided")
 	}
 
+	if artifactInfo.Type == ftypes.ArtifactFilesystem {
```
Comment (knqyf263) [inline pkg/scanner/scan.go:125]: I want to define artifact.Artifact.Clean(types.ArtifactReference) and call it in defer.

File: rpc/cache/service.proto [hunk 3]
```diff
@@ -59,3 +60,7 @@ message MissingBlobsResponse {
   bool            missing_artifact = 1;
   repeated string missing_blob_ids = 2;
 }
+
+message DeleteBlobRequest {
+  string          blob_id = 1;
```
Comment (knqyf263) [inline rpc/cache/service.proto:65]: I realized `blob_ids` may be better to reduce RPC calls.

File: pkg/scanner/scan.go (hunk 2 above)
Comment (afdesk) [reply pkg/scanner/scan.go]: done

File: pkg/scanner/scan.go (hunk 1 above)
Comment (afdesk) [reply pkg/scanner/scan.go]: sure, removed.

File: rpc/cache/service.proto (hunk 3 above)
Comment (afdesk) [reply rpc/cache/service.proto]: done

File: pkg/cache/remote.go [hunk 4]
```diff
@@ -61,3 +61,12 @@ func (c RemoteCache) MissingBlobs(imageID string, layerIDs []string) (bool, []st
 	}
 	return layers.MissingArtifact, layers.MissingBlobIds, nil
 }
+
+// DeleteBlobs removes blobs by IDs from RemoteCache
+func (c RemoteCache) DeleteBlobs(blobIDs []string) error {
+	_, err := c.client.DeleteBlobs(c.ctx, rpc.ConvertToDeleteBlobsRequest(blobIDs))
+	if err != nil {
+		return xerrors.Errorf("unable to store cache on the server: %w", err)
```
Comment (knqyf263) [inline pkg/cache/remote.go:69]:
```suggestion
		return xerrors.Errorf("unable to delete blobs on the server: %w", err)
```

File: pkg/scanner/scan.go [hunk 5]
```diff
@@ -110,6 +110,11 @@ func (s Scanner) ScanArtifact(ctx context.Context, options types.ScanOptions) (t
 	if err != nil {
 		return types.Report{}, xerrors.Errorf("failed analysis: %w", err)
 	}
+	defer func() {
+		if err := s.artifact.Clean(artifactInfo); err != nil {
+			log.Logger.Warnf("The artifact %q wasn't removed: %v", artifactInfo.Name, err)
```
Comment (knqyf263) [inline pkg/scanner/scan.go:115]: We may not remove the artifact.

```suggestion
			log.Logger.Warnf("Failed to clean the artifact %q: %v", artifactInfo.Name, err)
```

File: pkg/cache/remote.go (hunk 4 above)
Comment (afdesk) [reply pkg/cache/remote.go]: updated

File: pkg/scanner/scan.go (hunk 5 above)
Comment (afdesk) [reply pkg/scanner/scan.go]: updated

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**